kythe/docs/rfc/2909.md

```diff
@@ -0,0 +1,145 @@
+---
+title: Kythe Columnar Serving Format
+date: 2018/08/08
+link: https://github.com/google/kythe/pulls/2909
+---
+
+# Background
+
+The current Kythe serving data format is just a simple key-value store from string keys (usually [Kythe tickets](https://kythe.io/docs/kythe-uri-spec.html)) to ProtocolBuffer message.  This format requires the Kythe PostProcessor to perform a complex reduction over a potentially massive number of decorations/references/edges/etc. to produce a single lookup value per node.  A set of CrossReferences can be so large that it already requires manual paging, another complex transformation in the PostProcessor.  These operations tend to create hot shards and stragglers, leading to long post-processing times and potentially failures due to memory exhaustion.  Likewise, the server is forced to slowly decode large ProtocolBuffer messages even when only a subset of the data is required and the large value sizes can cause issues with some backend stores.
+
+# Proposal
+
+Split serving table ProtocolBuffer values into their component fields, making each an independent key-value entry.  Single lookup API calls become scans over a key prefix; paging becomes trivial.  Due the characteristics of LevelDB/SSTables, the split key-value entries will most likely be read together (see SSTable block encoding), keys will be prefix-encoded, and there should be no significant performance hit for reads.  On the contrary, since the Kythe API gives users the ability to request partial data, the split key-value entries allows the server to possibly skip over large segments of data while scanning and save the cost of decoding for a performance improvement.  Examples include a user requesting only small span of a file's decorations or a user requesting only a single kind of cross-references.
+
+More importantly, keeping serving data separate lets the Kythe PostProcessor do fewer large joins.  There will no longer need to be a join of all file decorations to create the single `FileDecorations` proto (likewise for CrossReferences).  Fewer large joins avoids stragglers in the PostProcessor and removes the need for many "limiters" (i.e. cutoff values for the size of outputs).
+
+The scope of this proposal only includes FileDecorations and CrossReferences.  Other services could be made to use similar formats, but the most improvement will be gained from these two APIs.
```

Comment:
I think this is more about the scope of the _implementation_ rather than the proposal—the proposal is actually nicely unspecific to these particular message types.

Reply:
✓